doc/adr/0002-nats-typed-messages.md

```diff
@@ -0,0 +1,182 @@
+# 2. NATS Typed Messages
+
+Date: 2020-05-06
+
+## Status
+
+Accepted
+
+## Context
+
+NATS Server has a number of JSON based messages - monitoring, JetStream API and more. These are consumed,
+and in the case of the API produced, by 3rd party systems in many languages. To assist with standardization
+of data validation, variable names and more we want to create JSON Schema documents for all our outward facing
+JSON based communication. Specifically this is not for server to server communication protocols.
+
+This effort is ultimately not for our own use - though libraries like `jsm.go` will use these to do validation
+of inputs - this is about easing interoperability with other systems and to eventually create a Schema Registry.
+
+There are a number of emerging formats for describing message content:
+
+ * JSON Schema - transport agnostic way of describing the shape of JSON documents
+ * AsyncAPI - middleware specific API description that uses JSON Schema for payload descriptions
+ * CloudEvents - standard for wrapping system specific events in a generic, routable, package. Supported by all
+   major Public Clouds and many event gateways. Can reference JSON Schema.
+ * Swagger / OpenAPI - standard for describing web services that uses JSON Schema for payload descriptions
+
+In all of these many of the actual detail like how to label types of event or how to version them are left up
+to individual projects to solve. This ADR describes how we are approaching this.
+
+## Decision
+
+### Overview
+
+We will start by documenting our data types using JSON Schema Draft 7. AsyncAPI and Swagger can both reference
+these documents using remote references so this, as a starting point, gives us most flexibility and interoperability
+to later create API and Transport specific schemas that reference these.
+
+We define 2 major type of typed message:
+
+ * `Message` - any message with a compatible `type` hint embedded in it
+ * `Event` - a specialized `message` that has timestamps and event IDs, suitable for transformation to 
+   Cloud Events. Typically, published unsolicited.
+
+Today NATS Server do not support publishing Cloud Events natively however a bridge can be created to publish
+those to other cloud systems using the `jsm.go` package that supports converting `events` into Cloud Event format.
+
+### Message Types
+
+There is no standard way to indicate the schema of a specific message. We looked at a lot of prior art from CNCF
+projects, public clouds and more but found very little commonality. The nearest standard is the Uniform Resource Name
+which still leaves most of the details up to the project and does not conventionally support versioning.
+
+We chose a message type like `io.nats.jetstream.api.v1.consumer_delete_response`, `io.nats.server.advisory.v1.client_connect`
```

Comment:
so one thought is that while the URN notation has always been a bit weird, there's one thing here that maybe would be helpful to normalize. Currently, we have `io.nats.jetstream.api` for the source, but perhaps we could standardize on a standard source like `io.nats.jetstream-api` that would make it easier to locate versions much in the same way we are collapsing the actual type to one of the tokens. Perhaps this also flattens the entire set of tokens to 5.